teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,609 @@
+
+### Simplicity
+
+`Conditions` actually manifest themselves as `Tasks` but are implemented as a separate CRD which makes them complex. Maintaining the separate Condition CRD takes extra and unnecessary effort, given that it's really a `Task` underneath. We prefer to reuse existing components when possible. 
+
+**Efficiency**
+
+Checking `Conditions` is slow and expensive because it spins up new `Pods` to evaluate each `Condition`. For example, Tekton dogfooding has been heavily using `Conditions` to decide what to run e.g only run this branch if it contains a specific type of file. These small checks add up such that many pods are used to check `Conditions` and it becomes slow and expensive. Even worse, we don't have a conditional branching construct (if/else or switch), so users have to implement and execute opposite `Conditions` which makes it even slower and more expensive. This can also be a problem in terms of the resource limits, requests, quotas and LimitRanges. 
+
+**Skipping**
+
+When a `Condition` fails, the guarded `Task` and its branch (dependent `Tasks`) are skipped. A `Task` is dependent on and in the branch of another `Task` as specified by ordering using `runAfter` or by resources using `Results`, `Workspaces` and `Resources`.  In some use cases of `Conditions`, when a `Condition` evaluates to `False`, users need to skip the guarded `Task` only and allow dependent `Tasks` to execute. An example use case is when there’s a particular `Task` that a Pipeline wants to execute when the git branch is dev/staging/qa, but not when it’s the main/master/prod branch. Another use case is when a user wants to send out a notification whether or not a parent guarded `Task` was executed, as described in [this issue](https://github.com/tektoncd/pipeline/issues/2937). 
+
+**Status**
+    
+It is currently difficult to distinguish between a `Task` that was skipped due to `Condition` evaluating as `False` and a `Task` that failed. This is because we use exit codes in `Check` as described in [Conditions](#conditions) section above and reuse `ConditionSucceeded` from Knative which can have the following states: `True`, `False` or `Unknown`. When a `Task` either fails or is skipped from `Condition` evaluating to `False`, we mark `ConditionSucceeded` as `False`. 
+
+## Requirements
+
+1. Users need a way to specify logic - `Guards` - to determine if a `Task` should execute or not.
+1. Users need a way to specify whether to execute dependent `Tasks` when `Guards` of a parent `Task` evaluate to `False`.
+1. Users should be able to specify multiple `Guards` for a given `Task`.
+1. Users should be able to use `Outputs` from parent `Tasks` and static input to specify `Guards`.
```

Comment:
* What happens when one of the `Guards` fail? Do we exit the pipeline just like its done today?

* How are we addressing `Guard` not producing any task result? The way its works with `DAG` is: A `DAG` task could exit successfully without instantiating a task result `/tekton/results/foo`. The pipeline exits with failure if this result is referenced by any other task since the result file does not exist. We could follow the same behavior. 🤔

Reply:
>What happens when one of the Guards fail? Do we exit the pipeline just like its done today?

That is not how it works today. If a guarded tasks condition evaluates to _false_  that task is omitted, and the tasks that depend on it is omitted - but the Pipeline continue to run to completion.

This was a bit unclear from docs before, but I tried to expand on that in https://github.com/tektoncd/pipeline/pull/2636